### PR TITLE
Destructive tuple update

### PR DESCRIPTION
**Pull request by frej**
Destructive tuple update

This PR implements destructive update of tuples. This allows the VM to update tuples in-place when it is safe to do so, thus improving performance by both doing less copying but also by producing less garbage. Destructive tuple update requires support from both the compiler and the runtime.

The compiler support extends the alias analysis implemented to support private append of binaries to track the alias status of tuples and their elements. The augmented aliasing information is used to find `update_record` instructions where the input tuple is unique (no other reference to it exists) and where it dies at the time of the update, as such instructions can safely update the tuple in-place, they are given the hint `inplace`.

The runtime support is implemented by Björn Gustavsson (@bjorng) who has also done some [bechmarking](https://gist.github.com/bjorng/8ae09b36790a70689a58b160fc4c24a9). For large tuples, we can observe an up to 25% reduction in execution time.

### CODE REVIEW

**Comment by garazdawi** [comment]
@github-actions disable-cache

File: lib/compiler/src/beam_ssa_alias.erl [hunk 1]
```diff
@@ -1417,141 +1258,119 @@ aa_breadth_first([], [], _Seen, _FuncDb) ->
...
-    aa_assert_parent_of(V, Ps, SS);
-aa_assert_parent_of([V|Vs], Child, SS) ->
-    aa_assert_parent_of(V, Child, SS),
-    aa_assert_parent_of(Vs, Child, SS);
-aa_assert_parent_of(_, [], _) ->
-    true;
-aa_assert_parent_of([], _, _) ->
-    true.
+expand_record_update(#opt_st{ssa=Linear0,cnt=First,anno=Anno0}=OptSt) ->
+    {Linear,Cnt} = eur_blocks(Linear0, First),
```
[the remark this answers is not in the document]
Comment (frej) [reply lib/compiler/src/beam_ssa_alias.erl:1262]: Correct, too bad I can't claim dyslexia :)

File: lib/compiler/src/beam_ssa_alias.erl [hunk 2]
```diff
@@ -1162,49 +889,161 @@ aa_alias_inherit_and_alias_if_arg_does_not_die(Dst, Arg, SS0, AAS) ->
...
+            true;
+        #t_float{} ->
+            true;
+        'pid' ->
+            true;
+        'port' ->
+            true;
+        'reference' ->
+            true;
+        _ ->
```
[the remark this answers is not in the document]
Comment (frej) [reply lib/compiler/src/beam_ssa_alias.erl:936]: 'identifier' is missing, I'll add a commit to cover that case.

As for `nil`, my thinking is that it for most practical cases where we can benefit from destructive updates, it only occurs as the tail of a list. As the type analysis pass basically punts on the type of the tail of a pair and says that the type is `any()` there is no point in doing anything but to say --- Oh, it's a nil, that's a list (which it is according to the type lattice) and stop trying to do anything intelligent.

**Comment by frej** [comment]
I have pushed a fixup and an addition. As the fixup will trigger some conflicts, the end result is available in [frej:frej/destructive-tuple-update-merged-fixups](https://github.com/frej/otp/tree/frej/destructive-tuple-update-merged-fixups), with the preload and bootstrap removed. @bjorng, feel free to regenerate preload and bootstrap and then force-push the result to this PR ([frej:frej/destructive-tuple-update](https://github.com/frej/otp/tree/frej/destructive-tuple-update))

**Pull request merged**